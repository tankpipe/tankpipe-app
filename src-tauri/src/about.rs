use serde::{Serialize, Deserialize};
use data_encoding::BASE64;


const VERSION: &str = env!("CARGO_PKG_VERSION");
const CONTACT: &str = "aHR0cHM6Ly9kaXNjb3JkLmdnL3V6cTg5R1ByVWc=";

#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq  )]
pub struct About {
    pub version: String,
    pub contact: String,
}

impl About {
    pub fn new() -> About {
        let input: Vec<u8> = CONTACT.into();
        let binding = BASE64.decode(&input).unwrap();
        let s = String::from_utf8_lossy(&binding);
        About{
            version: VERSION.to_string(),
            contact: s.into_owned()
        }
    }
}
