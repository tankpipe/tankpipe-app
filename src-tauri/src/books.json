{   
    "accounts": {
      "44742ea3-f835-4267-b62e-0c1d4796d575": {
        "id": "44742ea3-f835-4267-b62e-0c1d4796d575",
        "name": "Savings Account 1",
        "account_type": "Debit",
        "balance": "0",
        "starting_balance": "0"
      },
      "66ac7995-4991-495a-b44e-d9882d579260": {
        "id": "66ac7995-4991-495a-b44e-d9882d579260",
        "name": "Credit Account 1",
        "account_type": "Credit",
        "balance": "0",
        "starting_balance": "0"
      }
    },
    "scheduled_transactions": [
      {
        "id": "8565ba58-2df8-4690-b4f1-7849b2b4b48f",
        "name": "Some income",
        "period": "Months",
        "frequency": 1,
        "start_date": "2022-06-04",
        "last_date": "2022-06-04",
        "amount": "200",
        "description": "Money in",
        "dr_account_id": "cb946148-79b0-43bc-a3b8-b8bd811aa1ad",
        "cr_account_id": "c77327ef-d27d-434c-a205-2e975e822c87"
      }
    ],
    "transactions": [
      {
        "id": "4ed6200e-4d96-4c40-b140-36de0cfc7bc8",
        "date": "2022-06-04",
        "description": "received moneys",
        "dr_account_id": "44742ea3-f835-4267-b62e-0c1d4796d575",
        "cr_account_id": "66ac7995-4991-495a-b44e-d9882d579260",
        "amount": "10000",
        "status": "Recorded"
      },
      {
        "id": "f70e1b5d-12ee-47b0-9fbb-ab7671ff4595",
        "date": "2022-06-05",
        "description": "Gave some moneys back",
        "dr_account_id": "66ac7995-4991-495a-b44e-d9882d579260",
        "cr_account_id": "44742ea3-f835-4267-b62e-0c1d4796d575",
        "amount": "98.99",
        "status": "Recorded"
      },
      {
        "id": "4ed6200e-4d96-4c40-b140-36de0cfc7bc9",
        "date": "2022-07-01",
        "description": "received more moneys",
        "dr_account_id": "44742ea3-f835-4267-b62e-0c1d4796d575",
        "cr_account_id": "66ac7995-4991-495a-b44e-d9882d579260",
        "amount": "250",
        "status": "Recorded"
      },
      {
        "id": "4ed6200e-4d96-4c40-b140-36de0cfc7bd1",
        "date": "2022-07-01",
        "description": "received more moneys",
        "dr_account_id": "44742ea3-f835-4267-b62e-0c1d4796d575",
        "cr_account_id": null,
        "amount": "150.45",
        "status": "Recorded"
      }
    ]
  }