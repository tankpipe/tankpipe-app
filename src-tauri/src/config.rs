use std::{ffi::{OsString, OsStr}, path::PathBuf};
use serde::{Deserializer, Serializer, Serialize, Deserialize};

#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq  )]
pub struct FileDetails {
    pub name: String,
    #[serde(serialize_with = "serialize_osstring")]
    #[serde(deserialize_with = "deserialize_osstring")]
    pub path: OsString,
}

impl FileDetails {
    pub fn empty() -> FileDetails {
        FileDetails{ name: "".to_string(), path: OsString::from("") }
    }

    pub fn from_path(name: &str, path: PathBuf) -> FileDetails {
        FileDetails{ name: name.to_string(), path: path.into_os_string() }
    }

    pub fn from_os_string(name: &str, path: OsString) -> FileDetails {
        FileDetails{ name: name.to_string(), path: path }
    }
}

#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq  )]
pub enum DateFormat {
    Locale,
    Regular,
    US,
    ISO,
}

#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq  )]
pub struct Config {
    pub version: String,
    #[serde(serialize_with = "serialize_osstring")]
    #[serde(deserialize_with = "deserialize_osstring")]
    pub data_dir: OsString,
    #[serde(serialize_with = "serialize_osstring")]
    #[serde(deserialize_with = "deserialize_osstring")]
    pub config_dir: OsString,
    pub last_file: FileDetails,
    pub recent_files: Vec<FileDetails>,
    pub display_date_format: DateFormat,
    pub import_date_format: String
}

impl Config {
    pub fn data_dir_as_os_str(&self) -> &OsStr {
        &self.data_dir.as_os_str()
    }

    pub fn settings_path(&self) -> PathBuf {
        PathBuf::from(self.config_dir.clone()).join("settings.json")
    }

    pub fn file_path(&self, name: &str) -> PathBuf {
        PathBuf::from(self.data_dir.clone()).join(name)
    }

    pub fn set_last_from_path(&mut self, path: OsString, name: &str) {
        let last_file: FileDetails;
        let os_string_path = OsString::from(path);
        self.last_file.path = os_string_path.clone();
        self.last_file.name = name.to_string();
        let index = self.recent_files.iter().position(|f| *f.path == os_string_path);
        match index {
            Some(i) => {
                last_file = self.recent_files.get(i).unwrap().clone();
                self.recent_files.remove(index.unwrap());
            },
            None =>last_file = FileDetails::from_os_string(name, os_string_path)
        }
        self.recent_files.insert(0, last_file);
    }

    pub fn set_last(&mut self, last_file: FileDetails) {
        if let Some(index) = self.recent_files.iter().position(|f| *f.path == last_file.path) {
            self.recent_files.remove(index);
        }
        self.last_file = last_file.clone();
        self.recent_files.insert(0, last_file);
    }

}

pub fn deserialize_osstring<'de, D>(deserializer: D) -> Result<OsString, D::Error>
    where D: Deserializer<'de>
{
    let path = String::deserialize(deserializer)?;
    Ok(OsString::from(path))
}

pub fn serialize_osstring<S>(path: &OsString, serializer: S) -> Result<S::Ok, S::Error>
    where S: Serializer,
{
    serializer.serialize_some(&PathBuf::from(path.as_os_str()))
}