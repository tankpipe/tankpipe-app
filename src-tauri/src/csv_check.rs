use serde::{Serialize, Deserialize};


use crate::reader::ColumnTypes;

#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct CsvCheck {
    pub column_types: ColumnTypes,
    pub header_row: Vec<String>,
    pub sample_rows: Vec<Vec<String>>,
    pub mapping_exists: bool,
}

impl CsvCheck {
    pub fn create_new(column_types: ColumnTypes, header_row: Vec<String>, sample_rows: Vec<Vec<String>>) -> CsvCheck {
        CsvCheck {
            column_types: column_types,
            header_row: header_row,
            sample_rows: sample_rows,
            mapping_exists: false,
        }
    }

    pub fn set_mapping_exists(self, mapping_exists: bool) -> CsvCheck {
        CsvCheck {
            column_types: self.column_types,
            header_row: self.header_row,
            sample_rows: self.sample_rows,
            mapping_exists: mapping_exists,
        }
    }
}


