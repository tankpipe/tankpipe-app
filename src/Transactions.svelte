<script>
	import EditTransaction from './EditTransaction.svelte'
    import Select from './Select.svelte'

	export let curAccount
	export let accounts = []
	let mode = "TRANSACTIONS"
	let editMode = "ADD"
	let curTransaction

	const close = () => {
        console.log("close")
        mode = "TRANSACTIONS";
		if (curAccount) loadTransactions();
    }

	$: {
		if (curAccount) loadTransactions();
    }

	const selectTransaction = (transaction) => {
		curTransaction = transaction
		editMode = "EDIT"
		mode = "EDIT"
	}

	let transactions = [];
	export const loadTransactions = async () => {
		console.log("loadTransactions: " + curAccount.id);
   		transactions = await invoke('transactions', {accountId: curAccount.id});
		console.log(transactions)
	};

	const formatter = new Intl.NumberFormat('en-AU', {
		minimumFractionDigits: 2,
		maximumFractionDigits: 2,
	});
	const getDebitAmount = (transaction, curAccount) => {
		return transaction.transaction_type === "Debit" ? formatter.format(transaction.amount) : ''
	}

	const getCreditAmount = (transaction, curAccount) => {
		return transaction.transaction_type === "Credit" ? formatter.format(transaction.amount) : ''
	}

	const getBalance = (transaction) => {
		return formatter.format(transaction.balance)
	}

	const handleAddClick = () => {
		editMode = "ADD"
		mode = "EDIT"
	}

</script>


<div class="account-heading">
	{#if mode === "TRANSACTIONS" && curAccount}
	<Select bind:item={curAccount} items={accounts}/>
	<div class="toolbar"><i class="gg-add-r" on:click="{handleAddClick(curAccount)}"></i></div>
	{/if}
</div>
{#if mode === "EDIT"}
<EditTransaction {close} {curAccount} {accounts} {editMode} {curTransaction} />
{/if}
{#if mode === "TRANSACTIONS"}
<div class="scroller">
	<table>
		<tr><th>Date</th><th>Description</th><th>Debit</th><th>Credit</th><th>Balance</th></tr>
		{#each transactions as t}
			<tr on:click={() => selectTransaction(t)}><!--{t.id}--><td>{t.date}</td><td class="description">{t.description}</td><td class="money">{getDebitAmount(t, curAccount)}</td><td class="money">{getCreditAmount(t, curAccount)}</td><td class="money">{getBalance(t)}</td></tr>
		{/each}
	</table>
</div>
{/if}

<style>
    .scroller{
		height: 100%;
		overflow: scroll;
	}
	td {
		text-align: left;
		overflow: hidden;
		line-height: 1em;
		color: #444;
		background-color: #f0f0f0;
		padding: 8px;
	}

	th {
		color:#f0f0f0;
		background-color: #444;
		font-weight: 400;
		font-size: .8em;
	}

	tr:hover td {
		cursor: pointer;
		background-color: #FFF;
	}

	.money {
		text-align: right !important;
		min-width: 100px;
		font-family: 'Courier New', Courier, monospace;
		font-weight: bold;
	}

	.description {
		min-width: 300px;
		font-size: 1.05em;
	}

	.account-heading {
		text-align: left;
	}

	.toolbar {
		float: right;
		color: #C0C0C0;
	}

	.toolbar i:hover{
		color: #F0F0F0;
		border-color: #f0f0f0;
	}

	.gg-add-r {
		box-sizing: border-box;
		position: relative;
		display: block;
		width: 22px;
		height: 22px;
		border: 2px solid;
		transform: scale(var(--ggs,1));
		border-radius: 4px
	}

	.gg-add-r::after,
	.gg-add-r::before {
		content: "";
		display: block;
		box-sizing: border-box;
		position: absolute;
		width: 10px;
		height: 2px;
		background: currentColor;
		border-radius: 5px;
		top: 8px;
		left: 4px
	}

	.gg-add-r::after {
		width: 2px;
		height: 10px;
		top: 4px;
		left: 8px
	}

</style>