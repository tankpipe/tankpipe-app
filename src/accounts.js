import { writable } from 'svelte/store'

const accounts = writable({})

export {accounts as accounts}