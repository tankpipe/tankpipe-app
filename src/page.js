import { writable } from 'svelte/store'

const views = {
    BOOKS: "BOOKS",
    ACCOUNTS: "ACCOUNTS",
    TRANSACTIONS: "TRANSACTIONS",
    SCHEDULES: "SCHEDULES",
    SETTINGS: "SETTINGS",
    NET_ASSETS: "NET_ASSETS",
    JOURNAL: "JOURNAL",
}

const modes = {
    NEW: "NEW",
    EDIT: "EDIT",
    MULTI_EDIT: "MULTI_EDIT",
    LIST: "LIST",
    LOAD: "LOAD"
}

const page = writable({view: views.ACCOUNTS, mode:modes.LIST, payload:{}})

const isEditMode = (page) => {
    return page.mode === modes.EDIT || page.mode === modes.NEW || page.mode === modes.MULTI_EDIT
}

const isSingleEditMode = (page) => {
    return page.mode === modes.EDIT || page.mode === modes.NEW
}

const isMultiEditMode = (page) => {
    return page.mode === modes.MULTI_EDIT
}

const isListMode = (page) => {
    return page.mode === modes.LIST
}

export {page, views, modes, isEditMode, isMultiEditMode, isSingleEditMode, isListMode}