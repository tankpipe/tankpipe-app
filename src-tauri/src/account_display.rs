use accounts::account::{AccountType, Account};
use chrono::NaiveDate;
use rust_decimal::Decimal;
use serde::{Serialize, Deserialize};
use uuid::Uuid;
use accounts::serializer::*;
use rust_decimal_macros::dec;

#[derive(Clone, Serialize, Deserialize)]
pub struct NewAccount {
	pub name: String,
	pub account_type: AccountType,
	pub starting_balance: Decimal
}

impl NewAccount {
	pub fn to_account(self) -> Account {
		Account {
			id: Uuid::new_v4(),
			name: self.name,
			starting_balance: self.starting_balance,
			account_type: self.account_type,
            balance: dec!(0)
		}
	}

}


#[derive(Clone, Serialize, Deserialize)]
pub struct DateParam {
    #[serde(serialize_with = "serialize_naivedate")]
    #[serde(deserialize_with = "deserialize_naivedate")]
	pub date: NaiveDate
}

