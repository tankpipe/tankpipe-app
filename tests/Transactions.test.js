import { render, waitFor } from '@testing-library/svelte'
import Transactions from '../src/Transactions.svelte'
import {accounts} from '../src/accounts.js'
import {page, views, modes} from '../src/page'
import account_data from './account_data.json'
import transaction_data from './transaction_data.json'

accounts.set(account_data)

it('is displayed correctly', async () => {
    page.set({view: views.TRANSACTIONS, mode: modes.LIST})
    let entries = [transaction_data[0].entries[0], transaction_data[1].entries[0]]
    const mockFetchTransactions = jest.fn(() => Promise.resolve(entries));
    global.invoke = mockFetchTransactions;

    const {findByTitle, container} = render(Transactions, {curAccount: account_data[0]})
    await waitFor(() => expect(mockFetchTransactions).toHaveBeenCalledTimes(1))
    const _waitForRenderUpdate = await findByTitle('Add a transaction')

    // Note: Select should show Account 1 as selected but defaults to Account 3
    expect(container.outerHTML).toMatchSnapshot();
});

