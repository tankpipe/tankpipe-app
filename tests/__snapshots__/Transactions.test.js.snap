// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`is displayed correctly 1`] = `"<body><div><div class=\\"account-heading\\"><div class=\\"widgetNoLabel svelte-vv6jek\\"> <select name=\\"itemSelect\\" class=\\"svelte-vv6jek flat\\"><option value=\\"[object Object]\\">Account 1</option><option value=\\"[object Object]\\">Account 2</option><option value=\\"[object Object]\\">Account 3</option></select></div> <div class=\\"toolbar svelte-1qkpc1v\\"><div class=\\"toolbar-icon import-icon svelte-1qkpc1v\\" title=\\"Import transactions\\"></div> <div class=\\"toolbar-icon svelte-1qkpc1v\\" title=\\"Add a transaction\\"></div></div></div>  <div class=\\"widget errors\\"> </div> <div class=\\"scroller svelte-1qkpc1v\\"><table class=\\"svelte-1qkpc1v\\"><tr><th class=\\"justify-left svelte-1qkpc1v\\">Date</th><th class=\\"justify-left svelte-1qkpc1v\\">Description</th><th class=\\"svelte-1qkpc1v\\">Debit</th><th class=\\"svelte-1qkpc1v\\">Credit</th><th class=\\"svelte-1qkpc1v\\">Balance</th></tr> <tr class=\\"svelte-1qkpc1v\\"><td class=\\"svelte-1qkpc1v\\">2023-01-21</td> <td title=\\"A test transaction\\" class=\\"svelte-1qkpc1v\\"><div class=\\"description svelte-1qkpc1v\\">A test transaction</div></td> <td class=\\"money svelte-1qkpc1v\\">200.00</td> <td class=\\"money svelte-1qkpc1v\\"></td> <td class=\\"money svelte-1qkpc1v\\">200.00</td> </tr><tr class=\\"svelte-1qkpc1v\\"><td class=\\"svelte-1qkpc1v\\">2023-01-21</td> <td title=\\"A test compound transaction\\" class=\\"svelte-1qkpc1v\\"><div class=\\"description svelte-1qkpc1v\\">A test compound transaction</div></td> <td class=\\"money svelte-1qkpc1v\\">200.00</td> <td class=\\"money svelte-1qkpc1v\\"></td> <td class=\\"money svelte-1qkpc1v\\">400.00</td> </tr></table> </div></div></body>"`;
